lib/auth/webauthn/login_test.go

```diff
@@ -621,6 +654,162 @@ func TestCredentialRPID(t *testing.T) {
+	const webUser = "llama"
+	webIdentity := newFakeIdentity(webUser)
+	webConfig := &types.Webauthn{RPID: "example.com"}
+
+	const webOrigin = "https://example.com"
+	ctx := context.Background()
+
+	// Register a Webauthn device.
+	// Last registration step creates the user webID and adds the new device to
+	// identity.
+	webKey, err := mocku2f.Create()
+	require.NoError(t, err)
+	webKey.PreferRPID = true // Webauthn-registered device
+	webRegistration := &wanlib.RegistrationFlow{
+		Webauthn: webConfig,
+		Identity: webIdentity,
+	}
+	cc, err := webRegistration.Begin(ctx, webUser, false /* passwordless */)
+	require.NoError(t, err)
+	ccr, err := webKey.SignCredentialCreation(webOrigin, cc)
+	require.NoError(t, err)
+	device, err := webRegistration.Finish(ctx, wanlib.RegisterResponse{
+		User:             webUser,
+		DeviceName:       "webauthn1",
+		CreationResponse: ccr,
+	})
+	require.NoError(t, err)
+
+	tests := []struct {
+		name          string
+		challengeExt  mfav1.ChallengeExtensions
+		wantBeginErr  error
+		requiredExt   mfav1.ChallengeExtensions
+		wantFinishErr error
+	}{
+		{
+			name: "NOK scope not satisfied",
+			challengeExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION,
+			},
+			requiredExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_LOGIN,
+			},
+			wantFinishErr: trace.AccessDenied("required scope %q is not satisfied by the given webauthn session with scope %q", mfav1.ChallengeScope_CHALLENGE_SCOPE_LOGIN, mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION),
+		}, {
+			// Old clients do not yet provide a scope, so we only enforce scope
+			// opportunistically during login finish.
+			// TODO(Joerger): DELETE IN v16.0.0 - change to NOK
+			name: "OK scope not specified",
+			challengeExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_UNSPECIFIED,
+			},
+			requiredExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION,
+			},
+		}, {
+			name: "OK scope not required",
+			challengeExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION,
+			},
+			requiredExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_UNSPECIFIED,
+			},
+		}, {
+			name: "OK required scope satisfied",
+			challengeExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION,
+			},
+			requiredExt: mfav1.ChallengeExtensions{
+				Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_ADMIN_ACTION,
+			},
+		}, {
+			name: "NOK reuse not allowed for scope",
+			challengeExt: mfav1.ChallengeExtensions{
+				Scope:      mfav1.ChallengeScope_CHALLENGE_SCOPE_LOGIN,
+				AllowReuse: mfav1.ChallengeAllowReuse_CHALLENGE_ALLOW_REUSE_YES,
+			},
+			wantBeginErr: trace.BadParameter("mfa challenges with scope %s cannot allow reuse", mfav1.ChallengeScope_CHALLENGE_SCOPE_LOGIN),
```

Comment:
Given that this is the only begin error, I would be tempted to take if off the table and simplify it accordingly.